ractor.c

```diff
@@ -1678,6 +1678,91 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
     rb_raise(rb_eRactorMovedError, "can not send any methods to a moved object");
 }
 
+/*
+ *  Document-class: Ractor::ClosedError
+ *
+ *  Raised when an attempt is made to take something from the Ractor's outgoing port,
+ *  but it is closed with Ractor#close_outgoing, or to send something to Ractor's
+ *  incoming port, and it is closed with Ractor#close_incoming, or an attempt to
+ *  send/take something with ractor which was already terminated.
+ *
+ *     r = Ractor.new { sleep(500) }
+ *     r.close_outgoing
+ *     r.take # Ractor::ClosedError
+ *
+ *  ClosedError is a descendant of StopIteration, so closing of the ractor will break
+ *  the loops without propagating the error:
+ *
+ *     r = Ractor.new { 3.times { puts "Received: " + receive } }
+ *
+ *     loop { r.send "test" }
+ *     puts "Continue successfully"
+ *
+ *  This will print:
+ *
+ *     Received: test
+ *     Received: test
+ *     Received: test
+ *     Continue successfully
+ */
+
+/*
+ *  Document-class: Ractor::RemoteError
+ *
+ *  Raised on attempt to Ractor#take if there were exception in the Ractor. Its +cause+ will
+ *  contain the original exception, and +ractor+ is the current ractor.
+ *
+ *     r = Ractor.new { raise "Something weird happened" }
+ *
+ *     begin
+ *       r.take
+ *     rescue => e
+ *       p e         # => #<Ractor::RemoteError: thrown by remote Ractor.>
+ *       p e.ractor  # => #<Ractor:#2 test.rb:1 terminated>
+ *       p e.cause   # => #<RuntimeError: Something weird happened>
+ *     end
+ *
+ */
+
+/*
+ *  Document-class: Ractor::MovedError
+ *
+ *  Raised on an attempt to access an object which was moved in Ractor#send or Ractor.yield.
+ *
+ *     r = Ractor.new { receive }
+ *
+ *     ary = [1, 2, 3]
+ *     r.send(ary, move: true)
+ *     ary.inspect
+ *     # Ractor::MovedError (can not send any methods to a moved object)
+ *
+ */
+
+/*
+ *  Document-class: Ractor::MovedObject
+ *
+ *  A special object which replaces any value that was moved to another ractor in Ractor#send
+ *  or Ractor.yield. Any attempt to access the object results in Ractor::MovedError.
+ *
```

Comment:
Good point. I was actually surprised by this when playing with examples.
BTW, do you know why the object is so severely limited, would it break something if it would have not frozen class (like even `nil` has), nice `#inspect` an so on?

Reply:
I believe the intent is to enforce that it is a programming error to refer to that value, but I'm not sure. It was clever of you to do `MovedObject === obj`, I didn't think of that.
Note that refinements work even though the class is frozen...

Reply:
I'll open an issue, because it might be helpful to actually have some basic introspection info like `inspect`.

Reply:
BTW, it seems `p` bypasses `#inspect` redefinition (and `#to_s` redefinition to format the message!):
```ruby
ary = []
Ractor.new{ sleep }.send(ary, move: true)
ary.inspect
# Ractor::MovedError (can not send any methods to a moved object)
p ary
# NoMethodError (undefined method `inspect' for #<Ractor::MovedObject:0x00005598b4093dd0>)
```

Reply:
Yes, I believe this is standard refinement behavior; methods defined outside the `using` can not see refined methods. It can be confusing because some syntax sugar does, e.g. `"...#{foo}..."` is like syntax sugar for `"...#{foo.to_s}..."`, but that's all I can think of. `[foo].join` will not see a refined `to_s`